<a href="https://colab.research.google.com/github/MonicaMaVi/Algortimos_de_optimizacion/blob/master/Seminario_Algoritmos.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Algoritmos de optimización - Seminario<br>
Nombre y Apellidos: Mónica Mayayo Vives  <br>
Url: github.com/MonicaMaVi/Algortimos_de_optimizacion/Seminario_Algoritmos.ipynb<br>
Colab: https://colab.research.google.com/github/MonicaMaVi/Algortimos_de_optimizacion/blob/master/Seminario_Algoritmos.ipynb
Problema: 3. Combinar cifras y operaciones

Problema 3. Combinar cifras y operaciones
• El problema consiste en analizar el siguiente problema y diseñar un algoritmo que lo resuelva.
• Disponemos de las 9 cifras del 1 al 9 (excluimos el cero) y de los 4 signos básicos de las
operaciones fundamentales: suma(+), resta(-), multiplicación(*) y división(/)
• Debemos combinarlos alternativamente sin repetir ninguno de ellos para obtener una
cantidad dada. Un ejemplo sería para obtener el 4:
4+2-6/3*1 = 4
• Debe analizarse el problema para encontrar todos los valores enteros posibles planteando las
siguientes cuestiones:
- ¿Qué valor máximo y mínimo se pueden obtener según las condiciones del problema?

Se puede obtener un mínimo negativo ya que tenemos la operación menos y se pueden obtener números bastante alejados del 0, tanto negativos como positivos ya que tenemos la operación de multiplicación. En el código de abajo se adjunta el cálculo del máximo y el mínimo.

- ¿Es posible encontrar todos los valores enteros posibles entre dicho mínimo y máximo?

En el rango de enteros [-69, 77] no hay ningún número entero que no aparezca, si probamos con los números que están en los extremos, como -69, -68... y 77, 76... (cambiando el 77 en el código adjuntado abajo por el número que queramos probar) aparecen, por lo que podemos intuir que cuanto más nos acercamos al centro del rango (probando con 0, 1, -1) más veces van a aparecer esos números como resultado, y probándolo en el código así es. Es más probable que más operaciones den un resultado más próximo al centro del rango, en este caso el cuatro, al tener operaciones de suma y resta, por lo que si los números del extremo están, se intuye que los números centrales también estarán. 

He calculado en python, en el código abajo adjuntado, los números enteros que aparecen entre el máximo y el mínimo, los he guardado en un vector, he quitado los repetidos, he ordenado el vector y lo he mostrado en el resultado del programa, por lo que se puede ver que efectivamente aparecen todos los números enteros entre el máximo y el mínimo.

Si calculáramos el mínimo y el máximo cogiendo números decimales (nos daría el mínimo un valor de -70.71 y el máximo un valor de 78.83), allí sí que encontramos que el 78 que está entre el máximo y el mínimo pero no aparecen como resultado de ninguna fórmula obtenida.

In [ ]:
import itertools

N=list(range(1, 10))
signos = ['+', '-', '*', '/']

enteros=[]
minimo = 0
maximo = 0

# Permutaciones sin repetición de los números del 1 al 9 cogidos del 5 en 5
permuNum = list(itertools.permutations(N, 5))
# Permutaciones sin repetición de todos los signos
permuSignos = list(itertools.permutations(signos))

# Calculamos todas las fórmulas posibles
for i in permuNum: 
    for j in permuSignos: 
        formula = f"{i[0]}{j[0]}{i[1]}{j[1]}{i[2]}{j[2]}{i[3]}{j[3]}{i[4]}"
        resultado = eval(formula)
        # Cogemos solo números enteros
        if resultado==int(resultado):
            # vamos a crear un vector de resultados enteros para encontrar 
            # todos los valores enteros posibles entre el mínimo y el máximo.
            enteros.append(resultado)
            sin_repetidos = list(set(enteros)) # Quitamos los repetidos
            sin_repetidos.sort() # Lo ordenamos
            #Aquí mostraremos los resultados y las fórmulas que sean igual al número que elijamos, en este caso el 77
            if resultado == 77:
                print(formula)
                print(resultado)
            minimo = min(resultado,minimo)
            maximo = max(resultado,maximo)   

# Opcional: imprimir la cantidad total de permutaciones
print(f"Total de permutaciones: {len(permuNum)*len(permuSignos)}")
# Mostramos mínimo y máximo
print("Mínimo:", minimo, "Máximo:", maximo)
# Mostramos valores enteros entre el mínimo y el máximo
print("Valores enteros entre el mínimo y el máximo", sin_repetidos)

7/1-2+8*9
77.0
7/1-2+9*8
77.0
7/1+8*9-2
77.0
7/1+9*8-2
77.0
7-2/1+8*9
77.0
7-2/1+9*8
77.0
7-2+8/1*9
77.0
7-2+8*9/1
77.0
7-2+9/1*8
77.0
7-2+9*8/1
77.0
7-4/2+8*9
77.0
7-4/2+9*8
77.0
7-6/3+8*9
77.0
7-6/3+9*8
77.0
7+8/1*9-2
77.0
7+8*9/1-2
77.0
7+8*9-2/1
77.0
7+8*9-4/2
77.0
7+8*9-6/3
77.0
7+9/1*8-2
77.0
7+9*8/1-2
77.0
7+9*8-2/1
77.0
7+9*8-4/2
77.0
7+9*8-6/3
77.0
8/1*9-2+7
77.0
8/1*9+7-2
77.0
8*9/1-2+7
77.0
8*9/1+7-2
77.0
8*9-2/1+7
77.0
8*9-2+7/1
77.0
8*9-4/2+7
77.0
8*9-6/3+7
77.0
8*9+7/1-2
77.0
8*9+7-2/1
77.0
8*9+7-4/2
77.0
8*9+7-6/3
77.0
9/1*8-2+7
77.0
9/1*8+7-2
77.0
9*8/1-2+7
77.0
9*8/1+7-2
77.0
9*8-2/1+7
77.0
9*8-2+7/1
77.0
9*8-4/2+7
77.0
9*8-6/3+7
77.0
9*8+7/1-2
77.0
9*8+7-2/1
77.0
9*8+7-4/2
77.0
9*8+7-6/3
77.0
Total de permutaciones: 362880
Mínimo: -69.0 Máximo: 77.0
Valores enteros entre el mínimo y el máximo [-69.0, -68.0, -67.0, -66.0, -65.0, -64.0, -63.0, -62.0, -61.0, -60.0, -59.0, -58.0, -57.0, -56.0, -55.0, -54.0, -53.0, -52.0, -51.0, -50.0, -49.0, -48.0, -47.0, -46.0, -45.0, -4

(*)¿Cuantas posibilidades hay sin tener en cuenta las restricciones?<br>

¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones.


(*)¿Cuantas posibilidades hay sin tener en cuenta las restricciones?<br>

El número total de posibilidades son las permutaciones de los números (15120) multiplicadas por las permutaciones de los signos (24) que dan un total de 362880. Las permutaciones de los nueve números cogidos de cinco en cinco nos dan una cantidad de 15120 ya que nos quedaría una operación de factorial de 9 en el numerador y en el denominador (9-5)!. Las permutaciones de los cuatro signos cogidos de cuatro en cuatro son 24 posibilidades, ya que sería 4! en el numerador y en el denominador (4-4)! = 0! que es 1. Por último multiplicaríamos 15120*24 lo que nos da los 362880 números totales.

¿Cuantas posibilidades hay teniendo en cuenta todas las restricciones?

Teniendo encuenta la restricción de que solo cogemos números enteros, sería difícil calcularlo matemáticamente, por lo que esa operación la he hecho programando el ejercicio en python (que adjunto abajo) y me da un total de 90000 posibilidades. Otra restricción que plantea el ejercicio sería mirar que la suma de esas fórmulas me dé un cierto número, por ejemplo el 4, también lo he calculado en el ejercicio en python que adjunto, se podría calcular para cualquier número cambiando en el código el 4 por cualquier otro número. 

In [2]:
import itertools

N=list(range(1, 10))
signos = ['+', '-', '*', '/']
cantidad = 0
cantidad4 = 0

# Permutaciones sin repetición de los números del 1 al 9 cogidos del 5 en 5
permuNum = list(itertools.permutations(N, 5))
# Permutaciones sin repetición de todos los signos
permuSignos = list(itertools.permutations(signos))

# Calculamos todas las fórmulas posibles
for i in permuNum: 
    for j in permuSignos: 
        formula = f"{i[0]}{j[0]}{i[1]}{j[1]}{i[2]}{j[2]}{i[3]}{j[3]}{i[4]}"
        resultado = eval(formula)
        # Cogemos solo números enteros
        if resultado==int(resultado):
            cantidad +=1 
            #Aquí mostraremos los resultados y las fórmulas que sean igual al número que elijamos, en este caso el 4
            if resultado == 4:
                cantidad4 +=1

# Opcional: imprimir la cantidad total de permutaciones
print(f"Número total de permutaciones: {len(permuNum)*len(permuSignos)}")
print(f"Número total de permutaciones que son números enteros:", cantidad)
print(f"Número total de permutaciones que suman 4:", cantidad4)


Número total de permutaciones: 362880
Número total de permutaciones que son números enteros: 90000
Número total de permutaciones que suman 4: 2112


Modelo para el espacio de mejorSoluciones<br>
(*) ¿Cual es la estructura de datos que mejor se adapta al problema? Argumentalo.(Es posible que hayas elegido una al principio y veas la necesidad de cambiar, arguentalo)


La estructura principal es un listado de datos enteros, en las primeras partes del proyecto no tengo hecha esta estructura porque calculaba el máximo y el mínimo a la vez que iba recorriendo los listados de los enteros y los signos, de cara a los último ejericios del proyecto he pensado en hacerla por varias razones, la primera es que el código queda más claro y fácil de ejecutar. La segunda es que para el cálculo de la complejidad también es más fácil de explicar, sino al estar el cálculo del máximo y el mínimo entre los dos bucles for, quedaba más confusa la complejidad. Y la tercera es que para realizar un algoritmo que mejore la complejidad del algortimo por fuerza bruta también la voy a necesitar.

Hablando de otras estructuras de datos del ejercicio, como la de la variable fórmula que es una estructura f-string (formatted string) para poder manejarla con la función eval, al principio elegí un vector para manejar la información de las fórmulas pero pronto me dí cuenta que no me servía para poder transformarlo en una fórmula operable con eval.

In [12]:
# Permutaciones sin repetición de los números del 1 al 9 cogidos del 5 en 5
permuNum = list(itertools.permutations(N, 5))
# Permutaciones sin repetición de todos los signos
permuSignos = list(itertools.permutations(signos))

#Generamos un listado con los resultados enteros de las fórmulas
def crear_listado(N, signos):
  resultados_enteros = []
  for i in permuNum: 
    for j in permuSignos: 
        formula = f"{i[0]}{j[0]}{i[1]}{j[1]}{i[2]}{j[2]}{i[3]}{j[3]}{i[4]}"
        resultado = eval(formula)
        # Cogemos solo números enteros
        if resultado==int(resultado):
           resultados_enteros.append(resultado)
  return resultados_enteros

print(crear_listado(N, signos))

[1.0, 1.0, -3.0, 5.0, 1.0, -2.0, 1.0, 1.0, 4.0, 1.0, -3.0, 0.0, 5.0, 2.0, -5.0, -2.0, 7.0, 4.0, -1.0, -6.0, -3.0, 3.0, 8.0, 5.0, 1.0, -7.0, -4.0, 1.0, 9.0, 6.0, -3.0, 5.0, 5.0, -3.0, 3.0, 1.0, -1.0, 3.0, 3.0, -1.0, 2.0, 0.0, 1.0, 1.0, 1.0, 1.0, 0.0, 2.0, -1.0, 3.0, 0.0, 4.0, 1.0, -5.0, 1.0, 7.0, -1.0, 5.0, -2.0, 6.0, -3.0, 7.0, -9.0, 11.0, -4.0, 8.0, 4.0, 0.0, 1.0, 3.0, 0.0, 4.0, -1.0, 5.0, -2.0, 6.0, -5.0, 1.0, 7.0, 1.0, 7.0, 1.0, -5.0, 1.0, 4.0, 0.0, 5.0, -1.0, -1.0, 3.0, 3.0, 1.0, -3.0, 5.0, 1.0, 3.0, 0.0, -4.0, 2.0, 6.0, 0.0, 4.0, -3.0, -5.0, 5.0, 7.0, -1.0, 5.0, 6.0, -2.0, 4.0, 0.0, 3.0, 1.0, 1.0, 3.0, 0.0, 4.0, -1.0, 2.0, 3.0, 0.0, 7.0, -3.0, -3.0, 0.0, 5.0, 2.0, 5.0, -1.0, 0.0, -4.0, -1.0, 2.0, 6.0, 3.0, 4.0, 0.0, -5.0, -2.0, 7.0, 4.0, 3.0, 1.0, -7.0, -4.0, 9.0, 6.0, 1.0, 3.0, -9.0, 6.0, 11.0, -4.0, 3.0, 1.0, 8.0, -4.0, 6.0, -2.0, -3.0, 5.0, 5.0, -1.0, 4.0, 0.0, 3.0, 1.0, -7.0, 9.0, -12.0, 14.0, -7.0, 9.0, -7.0, 9.0, 9.0, -7.0, 5.0, -1.0, -7.0, 9.0, 9.0, -7.0, 5.0, -1.0, -12.0, 

Según el modelo para el espacio de soluciones<br>
(*)¿Cual es la función objetivo?

(*)¿Es un problema de maximización o minimización?

En este caso se han marcado dos funciones objetivo, calcular el máximo y el mínimo de los resultados de las distintas fórmulas que se van creando, por lo que tenemos un problema tanto de maximización como de minimización, ya que podemos escribir la fórmula como una función, que en nuestro caso las incógnitas no serían solo los dígitos sino también los signos, y tenemos que encontrar el máximo y el mínimo de esa función. Adjunto abajo el código del cálculo del máximo y el mínimo.

In [18]:
# Calculamos el máximo y el mínimo del listado por fuerza bruta
def fuerza_bruta(N, signos):
    minimo = 0
    maximo = 0
    for n in crear_listado(N, signos):
        minimo = min(n,minimo)
        maximo = max(n,maximo)  
    return minimo, maximo 

# Mostramos el mínimo y máximo
print("Mínimo:", minimo, "Máximo:", maximo)

Mínimo: -69.0 Máximo: 77.0


Diseña un algoritmo para resolver el problema por fuerza bruta

(Código adjuntado abajo, es el mismo de las 2 celdas anteriores) Para el problema por fuerza bruta generamos primero 2 listas de tuplas, una de números, donde las tuplas serán de longitud 5 y otra de signos, que será un listado de tuplas de longitud 4. El problema se puede realizar con 2 for, uno para recorrer los números y otro para recorrer los signos, dentro de estos bucles iremos comprobando si el nuevo número entero calculado es menor que el mínimo o mayor que el máximo en cuyo caso lo estableceremos como mínimo o máximo, según corresponda.

In [14]:
import itertools

N=list(range(1, 10))
signos = ['+', '-', '*', '/']
minimo = 0
maximo = 0

# Permutaciones sin repetición de los números del 1 al 9 cogidos del 5 en 5
permuNum = list(itertools.permutations(N, 5))
# Permutaciones sin repetición de todos los signos
permuSignos = list(itertools.permutations(signos))

#Generamos un listado con los resultados enteros de las fórmulas
def crear_listado(N, signos):
  resultados_enteros = []
  for i in permuNum: 
    for j in permuSignos: 
        formula = f"{i[0]}{j[0]}{i[1]}{j[1]}{i[2]}{j[2]}{i[3]}{j[3]}{i[4]}"
        resultado = eval(formula)
        # Cogemos solo números enteros
        if resultado==int(resultado):
           resultados_enteros.append(resultado)
  return resultados_enteros

listado_enteros = crear_listado(N, signos)
# Calculamos el máximo y el mínimo del listado
for n in listado_enteros:
    minimo = min(n,minimo)
    maximo = max(n,maximo)   

# Mostramos el mínimo y máximo
print("Mínimo:", minimo, "Máximo:", maximo)

Mínimo: -69.0 Máximo: 77.0


Calcula la complejidad del algoritmo por fuerza bruta

Este problema tiene un número de soluciones de 362880 posibilidades que es igual a 9!. Esto nos sugiere que la complejidad del algoritmo va a ser de n!, las permutaciones posibles.

Por otro lado en este algoritmo se calculan todas las variaciones sin repetición de n elementos tomados de k en k es decir O(n!/(n-k)!), llamamos a esta función dos veces, una para los números y otra para los signos. Por lo que vemos otra manera de ver que el orden de complejidad del algoritmo por fuerza bruta es 0(n!).

(*)Diseña un algoritmo que mejore la complejidad del algortimo por fuerza bruta. Argumenta porque crees que mejora el algoritmo por fuerza bruta

Respuesta

In [ ]:
#Utilizamos el algoritmo de BÚSQUEDA LOCAL para buscar el mínimo
import random

longitud = len((listado_enteros))  # largo único

def genera_vecinas_min(mejor_solucion, indice):
  #Generador de soluciones vecinas: sumarle o restarle uno al índice
  for i in range(100):
    if indice == len(list(set(listado_enteros)))-1:
      vecina = listado_enteros[indice-1]
      indice = indice-1
    else:
      vecina = listado_enteros[indice+1]
      indice = indice+1
    # La guardamos si mejora las anteriores
    if vecina <= mejor_solucion:
      mejor_solucion = vecina
    #print("Hola vecina: ", mejor_solucion)
  return mejor_solucion

# Aquí vamos a ejecutar la búsqueda local 100 veces para comprobar cuántas iteraciones se necesitan para dar con el mínimo
def genera_100_min():
  for i in range(100):
      # Para un índice válido en la lista (0 a longitud-1)
      indice = random.randint(0, longitud - 1)
      solucion = listado_enteros[indice]
      print("Solucion Incial:" , solucion)

      nueva_solucion = genera_vecinas_min(solucion, indice)
      print("Solución Final", nueva_solucion)
      
      if nueva_solucion == minimo:
          break
      print("Número de ejecuciones:", i)

genera_100_min()

Solucion Incial: 7.0
Solución Final -67.0
Número de ejecuciones: 0
Solucion Incial: 37.0
Solución Final -67.0
Número de ejecuciones: 1
Solucion Incial: 52.0
Solución Final -64.0
Número de ejecuciones: 2
Solucion Incial: -28.0
Solución Final -42.0
Número de ejecuciones: 3
Solucion Incial: -8.0
Solución Final -48.0
Número de ejecuciones: 4
Solucion Incial: 28.0
Solución Final -67.0
Número de ejecuciones: 5
Solucion Incial: 17.0
Solución Final -13.0
Número de ejecuciones: 6
Solucion Incial: 4.0
Solución Final -47.0
Número de ejecuciones: 7
Solucion Incial: 5.0
Solución Final -64.0
Número de ejecuciones: 8
Solucion Incial: 4.0
Solución Final -45.0
Número de ejecuciones: 9
Solucion Incial: 43.0
Solución Final -61.0
Número de ejecuciones: 10
Solucion Incial: 10.0
Solución Final -15.0
Número de ejecuciones: 11
Solucion Incial: 10.0
Solución Final -60.0
Número de ejecuciones: 12
Solucion Incial: 30.0
Solución Final -52.0
Número de ejecuciones: 13
Solucion Incial: 20.0
Solución Final -40.0
Núme

In [ ]:
# Vamos a calcular el tiempo que le cuesta al ordenador calcular el mínimo por búsqueda local para compararlo con el de fuerza bruta
import time

# Para un índice válido en la lista (0 a longitud-1)
indice = random.randint(0, longitud - 1)
solucion = listado_enteros[indice]

start = time.time()
genera_vecinas_min(solucion, indice)
end = time.time()
print(end-start)

0.25012779235839844


In [ ]:
def fuerza_bruta(N, signos):
    minimo = 0
    maximo = 0

    # Permutaciones sin repetición de los números del 1 al 9 cogidos del 5 en 5
    permuNum = list(itertools.permutations(N, 5))
    # Permutaciones sin repetición de todos los signos
    permuSignos = list(itertools.permutations(signos))

    # Calculamos todas las fórmulas posibles
    for i in permuNum: 
        for j in permuSignos: 
            formula = f"{i[0]}{j[0]}{i[1]}{j[1]}{i[2]}{j[2]}{i[3]}{j[3]}{i[4]}"
            resultado = eval(formula)
            # Cogemos solo números enteros
            minimo = min(resultado,minimo)
           # maximo = max(resultado,maximo)   

# Ahora calculamos el tiempo que le cuesta al ordenador calcular el mínimo por fuerza bruta para compararlo con el de búsqueda local
start = time.time()
fuerza_bruta(N, signos)
end = time.time()
print(end-start)

2.1938912868499756


In [ ]:
#Utilizamos el algoritmo de BÚSQUEDA LOCAL para buscar el máximo
import random

longitud = len((listado_enteros))  # largo único

def genera_vecinas_max(mejor_solucion, indice):
  #Generador de soluciones vecinas: sumarle o restarle uno al índice
  for i in range(100):
    if indice == len(list(set(listado_enteros)))-1:
      vecina = listado_enteros[indice-1]
      indice = indice-1
    else:
      vecina = listado_enteros[indice+1]
      indice = indice+1
    # La guardamos si mejora las anteriores
    if vecina >= mejor_solucion:
      mejor_solucion = vecina
    #print("Hola vecina: ", mejor_solucion)
  return mejor_solucion

# Aquí vamos a ejecutar la búsqueda local 100 veces para comprobar cuántas iteraciones se necesitan para dar con el máximo
for i in range(100):
    # Para un índice válido en la lista (0 a longitud-1)
    indice = random.randint(0, longitud - 1)
    solucion = listado_enteros[indice]
    print("Solucion Incial:" , solucion)

    nueva_solucion = genera_vecinas_max(solucion, indice)
    print("Solución Final", nueva_solucion)

    if nueva_solucion == maximo:
        break
    print("Número de ejecuciones:", i)


Solucion Incial: 39.0
Solución Final 53.0
Número de ejecuciones: 0
Solucion Incial: 33.0
Solución Final 71.0
Número de ejecuciones: 1
Solucion Incial: 38.0
Solución Final 61.0
Número de ejecuciones: 2
Solucion Incial: 66.0
Solución Final 75.0
Número de ejecuciones: 3
Solucion Incial: 25.0
Solución Final 49.0
Número de ejecuciones: 4
Solucion Incial: -32.0
Solución Final 69.0
Número de ejecuciones: 5
Solucion Incial: 7.0
Solución Final 24.0
Número de ejecuciones: 6
Solucion Incial: 24.0
Solución Final 69.0
Número de ejecuciones: 7
Solucion Incial: 17.0
Solución Final 54.0
Número de ejecuciones: 8
Solucion Incial: 9.0
Solución Final 50.0
Número de ejecuciones: 9
Solucion Incial: 27.0
Solución Final 75.0
Número de ejecuciones: 10
Solucion Incial: 14.0
Solución Final 44.0
Número de ejecuciones: 11
Solucion Incial: 42.0
Solución Final 76.0
Número de ejecuciones: 12
Solucion Incial: 7.0
Solución Final 71.0
Número de ejecuciones: 13
Solucion Incial: 6.0
Solución Final 59.0
Número de ejecucion

Como podemos ver el algoritmo de BÚSQUEDA LOCAL es más rápido que el de FUERZA BRUTA en tiempos pero para alcanzar el mismo valor que por fuerza bruta tenemos que ejecutarlo unas 60 veces más o menos, es decir se alcanza una vez de cada 60.

Si bien es cierto que no se suele quedar muy lejos del máximo, como podemos ver en los valores, algunos de los que nos aparecen sí que son bajos como el 24, 49 y 53... que estos sí quedan lejos del máximo, pero la mayoría de las veces nos quedamos muy próximos al máximo, lo mismo ocurre con el mínimo, algunos sí que son altos pero la mayoría son próximos al mínimo.

(*)Calcula la complejidad del algoritmo

La complejidad del algoritmo se ha reducido bastante, con unas 100 iteraciones la mayoría de las veces nos aproximamos bastante al mínimo y al máximo, eso es debido a que los números se repiten mucho dentro del listado. Hemos pasado de una complejidad O(n) a una complejidad O(1), ya que tengamos los elementos que tengamos siempre iteraremos 100 veces (en este caso).

Según el problema (y tenga sentido), diseña un juego de datos de entrada aleatorios

Respuesta

Aplica el algoritmo al juego de datos generado

Respuesta

Enumera las referencias que has utilizado(si ha sido necesario) para llevar a cabo el trabajo

Respuesta

Describe brevemente las lineas de como crees que es posible avanzar en el estudio del problema. Ten en cuenta incluso posibles variaciones del problema y/o variaciones al alza del tamaño

Respuesta